### ISSUE TITLE
regression: rustdoc resolution changes

### ISSUE THREAD

**Issue by Mark-Simulacrum** [rust-lang/rust#61560]
bint-0.1.0: [start](https://crater-reports.s3.amazonaws.com/beta-1.36-1/1.35.0/reg/bint-0.1.0/log.txt) v. [end](https://crater-reports.s3.amazonaws.com/beta-1.36-1/beta-2019-05-30/reg/bint-0.1.0/log.txt); cc @folkengine

Looks like resolution for the crate itself changed in some way, not sure.

**Comment by Mark-Simulacrum** [comment]
* lv2rs-core-0.3.0: [start](https://crater-reports.s3.amazonaws.com/beta-1.36-1/1.35.0/reg/lv2rs-core-0.3.0/log.txt) v. [end](https://crater-reports.s3.amazonaws.com/beta-1.36-1/beta-2019-05-30/reg/lv2rs-core-0.3.0/log.txt); cc @Janonard
 * oysterpack_testing-0.1.2: [start](https://crater-reports.s3.amazonaws.com/beta-1.36-1/1.35.0/reg/oysterpack_testing-0.1.2/log.txt) v. [end](https://crater-reports.s3.amazonaws.com/beta-1.36-1/beta-2019-05-30/reg/oysterpack_testing-0.1.2/log.txt); cc @oysterpack (maybe this issue, couldn't test due to openssl v0.9)

**Comment by Mark-Simulacrum** [comment]
* test-generator-0.2.2: [start](https://crater-reports.s3.amazonaws.com/beta-1.36-1/1.35.0/reg/test-generator-0.2.2/log.txt) v. [end](https://crater-reports.s3.amazonaws.com/beta-1.36-1/beta-2019-05-30/reg/test-generator-0.2.2/log.txt); cc @frehberg
 * trimmer-0.3.6: [start](https://crater-reports.s3.amazonaws.com/beta-1.36-1/1.35.0/reg/trimmer-0.3.6/log.txt) v. [end](https://crater-reports.s3.amazonaws.com/beta-1.36-1/beta-2019-05-30/reg/trimmer-0.3.6/log.txt); cc @tailhook

**Comment by pnkfelix** [comment]
triage: P-high.

**Comment by Janonard** [comment]
What does issue say? Do I as a crate owner have to worry about something?

**Comment by Mark-Simulacrum** [comment]
Presumably, the crates listed here are broken on current stable (1.36+) but this has not been verified. Someone will probably investigate soon and assign this to someone concrete to resolve.

**Comment by frehberg** [comment]
I will investigate my crate test-generator

**Comment by frehberg** [comment]
as for the test-generator, I tested various versions, cant see any error as in the log. There is no stdout for the error-case, any idea why it might fail?

**Comment by Janonard** [comment]
Btw, I found the problem with `lv2rs-core` and already fixed and updated it!

**Comment by nagisa** [comment]
Assigning to self to bisect the exact change which caused this regression, but it otherwise looks more like a rustdoc issue than a rustc one.

**Comment by pnkfelix** [comment]
hey @nagisa , do you think you will get around the bisection at some point? Or shall I assign to someone else?

**Comment by pnkfelix** [comment]
triage: reassigning to self.

**Comment by pnkfelix** [comment]
nominating for reassignment (want to get things that can be delegated off my plate)

**Comment by lqd** [comment]
I can repro on `lv2rs-core`, but looking at the logs, this was interesting:

Before:
```
Doc-tests bint
[INFO] [stdout] running 0 tests
```

After:
```
Doc-tests bint
[INFO] [stdout] running 1 test
```

This suggests a rustdoc change that detects more tests rather than resolution changes. 

Started failing in `1.36.0`.
Bisected to `nightly-2019-04-24`, and commit 0550766699a6602a51e361e8cb2825b540b7cce8 which is an update to rustdoc's markdown parser.

In this case, I think the old tests did not compile, but that wasn't easy to see until this version of rustdoc actually started testing them.

**Comment by lqd** [comment]
Looking around similar issues, there's #61562 and [this comment](https://github.com/rust-lang/rust/issues/61562#issuecomment-499643833) mentioning the same theory.

Puzzling, the code in the specific `lv2rs-core` doctest may have been ... correct ? `cargo test` still fails on 1.37.0 but passes on 1.38.0, or the latest nightly for example.

@pnkfelix what do you want to do with this issue ?

**Comment by pnkfelix** [comment]
@lqd it sounds to me like something we can either  hand off to the T-rustdoc team, or simply close outright.

**Comment by lqd** [comment]
OK, I'll keep bisecting the other crates, and also check if the failures reproduce on recent stables/nightlies, to see if we need to open more specific issues, and/or hand this one to T-rustdoc.

**Comment by lqd** [comment]
Update:
- `oysterpack_testing-0.1.2`, reproduces on the `beta-2019-05-30` tested in this crater run, but does not reproduce for the released 1.36 (or nightly), so it seems already fixed.
- `test-generator-0.2.2`, the setup seems complicated, there are proc_macros to help writing tests. These doctests rely on the unstable `#[bench]` feature, and was accepted e.g. on 1.35. It fails on `beta-2019-05-30`, works on 1.36 and 1.37, fails since 1.38 (and nightly) and is tracked in #64266. (It's interesting that it seems supposed to be a lint but via doctests it actually outputs an error).
It's likely that the previous rustdoc didn't see these doctests in this case as well. The interaction between doctests and this lint may require tuning ? I'll comment on the tracking issue.
- `trimmer-0.3.6`: works on 1.35, fails on this `beta-2019-05-30`, and works ever since 1.36, so it's likely close to the now-fixed issue`lv2rs-core` also had.

As a summary, it seems we have a mix of real rustc bugs, rustdoc effectively-breaking-changes-for-the-better testing new code, and invalid tests. The rustc bugs seem fixed, and the rustdoc behaviour change is #61562, and there might be miscommunication between rustdoc and rustc for the `soft_unstable` lint.

If need be, I could extract the doctests and bisect those to rustc commits, but I'm not sure we need to ?

**Comment by lqd** [comment]
As summarized earlier, this issue collects other issues, which were at the intersection of 
- the rustdoc behaviour change from 0550766699a6602a51e361e8cb2825b540b7cce8 and tracked in #61562
- and other rustc problems (possibly related to resolution) which have since been fixed, or are already tracked elsewhere (eg `#[bench]` accepted on stable).

While we could bisect _those_ problems to look for possible missing test cases, it's likely to point at known issues since they were all fixed either in previous betas of the same crater run or 2-3 releases ago, it doesn't seem absolutely necessary to do so. So let's close this one.
